server/src/main/kotlin/it/polito/wa2/group03/server/service/TicketingServiceStateful.kt

```diff
@@ -5,48 +5,52 @@ import io.jsonwebtoken.Jwts
 import it.polito.wa2.group03.server.model.TicketPayload
 import org.apache.tomcat.util.codec.binary.Base64
 import org.springframework.beans.factory.annotation.Value
+import org.springframework.context.annotation.Primary
 import org.springframework.stereotype.Service
 import java.util.*
 import java.util.concurrent.ConcurrentLinkedQueue
 
 @Service
-class TicketingServiceStateful(@Value("\${jwt.key}") private val key: String): TicketingServiceStateless(key) {
+@Primary
+class TicketingServiceStateful(@Value("\${jwt.key}") private val key: String) : ITicketingService {
 
     private var ticketQueue: Queue<String> = ConcurrentLinkedQueue()
     private val parser: JwtParser =
         Jwts.parserBuilder().setSigningKey(Base64.encodeBase64String(key.toByteArray())).build()
 
     override fun validateTicket(ticket: TicketPayload): ValidationResult {
 
-        try {
-
-            val sub = this.getSub(ticket.token)
-
-            when (sub in ticketQueue) {
-                true -> return ValidationResult.DUPLICATE
-                false -> ticketQueue.add(sub)
-            }
-
-            /**
-             * once we are sure the ticket is not a duplicate
-             * we can process it as we would do for any other.
-             */
-            return super.validateTicket(ticket)
+        val validation = super.validateTicket(ticket)
```

Comment:
while the rest looks good to me, I think this check should be performed after checking the `sub` field: if the request was already processed you would be processing it twice anyway, which is not ideal from a scalability standpoint.

every request that was already handled should not count towards server performance imo.

Reply:
I think once the JWT is parsed then the most expensive operation is likely done, and all those binaries checks in the interface are okish to re-do, especially considering the scope of this project. since you correctly mention that to get `sub` we must parse anyway, I am sold.